### PR TITLE
Expand COGETSERVERPID_OBJREFHDR struct and add type & clamping checks

### PR DESCRIPTION
**Pull request by forderud**
Done to ease debugging of "OBJREF" packages, including process ID's exceeding 65k.

### CODE REVIEW

**Comment by forderud** [comment]
Gentle reminder.

**Comment by kimgr** [comment]
Thank you! My only concern is that I would like for `cogetserverpid.h` to be valid C, to make it possible to use both from C and C++ (and any language with C bindings, really).

So if you could dumb it down a bit (no member functions, no `static_assert` [which is making its way into C, but only in very recent standards modes] and keep struct typedefs and declarations before statements) that would maintain that quality.

I'm getting my hopes up that there's a follow-up branch to fix this problem...? :smile: Or is this just to signal to callers that a clamped PID is not the real PID?

**Comment by forderud** [comment]
> Thank you! My only concern is that I would like for cogetserverpid.h to be valid C, to make it possible to use both from C and C++ (and any language with C bindings, really).

The current `cogetserverpid.h` implementation already performs C++ method invocations to `QueryInterface`, `Seek` & `Release`. Therefore, C compatibility is  already broken. The relevant  question then wold probably be whether you think it's ok to use C++11/14/17 constructs?

> I'm getting my hopes up that there's a follow-up branch to fix this problem...? 😄 Or is this just to signal to callers that a clamped PID is not the real PID?

Unfortunately, the PR will only detect that the PID is clamped. I don't yet know how to detect PID's above 65k.

**Comment by kimgr** [comment]
> The current cogetserverpid.h implementation already performs C++ method invocations

Ah, right you are. I thought I went through the trouble of C compatibility, but apparently not all the way.

That said, I'd rather keep the simple stuff C-friendly and come back to this existing problem later. So please mimic the code around you. And I'd rather avoid new-fangled C++, because it limits portability (this code is plain and simple enough that I don't want it to impose toolchain requirements on whoever embeds it.)

> Unfortunately, the PR will only detect that the PID is clamped. I don't yet know how to
> detect PID's above 65k.

OK, thanks. Then I wonder -- if the PID happens to be exactly 65535, it hasn't really been clamped and we'll reject a valid case. I'm not sure we should treat this as an error condition at all, but rather document it as a known limitation for now. Thoughts?

**Comment by forderud** [comment]
> That said, I'd rather keep the simple stuff C-friendly and come back to this existing problem later. So please mimic the code around you. And I'd rather avoid new-fangled C++, because it limits portability (this code is plain and simple enough that I don't want it to impose toolchain requirements on whoever embeds it.)

The only new C++ construct introduced in the PR is a `static_assert` (from C++11) check to prevent accidental in-memory padding. Everything else is C++98 compatible. Do you then want me to remove the static_assert?

> OK, thanks. Then I wonder -- if the PID happens to be exactly 65535, it hasn't really been clamped and we'll reject a valid case. I'm not sure we should treat this as an error condition at all, but rather document it as a known limitation for now. Thoughts?

You're correct that the case of PID=65535 will also be treated as if it was clamped and rejected. I personally like this behavior, since it's impossible for the calling code to detect if the PID then is actually 65535 or just a clamped value. Hence, the value cannot be trusted.

**Comment by kimgr** [comment]
Yes, I would like to remove the `static_assert`, and to keep the code at the same level of C-ness as it was (i.e. no member functions in the structs, keep the struct typedefs, etc.).

I'm trying to think what it means for the PID to be clamped. On the one hand, that's the PID value we parsed out of the payload, so everything went well. On the other hand, it indicates that we don't understand the protocol, and the real PID could be encoded some other way.

I was about to argue that I don't think it's `CoGetServerPID`'s responsibility to reject a bad PID, but if it indicates we don't understand the protocol, it's probably a good idea.

I don't think the refactoring changes add any significant value, so can you see if you can introduce the check for the right flags and the check for a clamped PID without redefining all the structs?

Thanks!

**Comment by forderud** [comment]
I've now tried to update the PR based on your proposal. Looking better now?

**Comment by forderud** [comment]
Gentle reminder.

**Comment by kimgr** [comment]
I'm out and about, I'll come back to this next week.

**Comment by kimgr** [comment]
Hey,

I did some other small whitespace cleanups on master, and then followed up with my interpretation of your changes: https://github.com/kimgr/cogetserverpid/pull/4

Looks good?

**Comment by forderud** [comment]
Closing the PR, since it's been replaced by https://github.com/kimgr/cogetserverpid/pull/4
